    accept_filter: Fix filter parameter handling
    
    For filters which implement accf_create, the setsockopt(2) handler
    caches the filter name in the socket, but it also incorrectly frees the
    buffer containing the copy, leaving a dangling pointer.  Note that no
    accept filters provided in the base system are susceptible to this, as
    they don't implement accf_create.
    
    Reported by:    Alexey Kulaev <alex.qart@gmail.com>
    Discussed with: emaste
    Security:       kernel use-after-free
    MFC after:      3 days
    Sponsored by:   The FreeBSD Foundation
    
    (cherry picked from commit 653a437c04440495cd8e7712c7cf39444f26f1ee)
    (cherry picked from commit 6008a5fad3c110c4ec03cc3fe60ce41c4e548b98)
--- sys/kern/uipc_accf.c.orig
+++ sys/kern/uipc_accf.c
@@ -298,6 +298,7 @@
 	so->sol_accept_filter = afp;
 	so->sol_accept_filter_arg = accept_filter_arg;
 	so->sol_accept_filter_str = accept_filter_str;
+	accept_filter_str = NULL;
 	so->so_options |= SO_ACCEPTFILTER;
 out:
 	SOCK_UNLOCK(so);
